TMEDIA 252 nav list item transparency issue

JackHowa opened the pull request:
## Description
When the sections menu has more items than can fit within the viewport, and the sections menu is closed additional white space is added to the bottom of the page.

## Jira Ticket
- [TMEDIA-252](https://arcpublishing.atlassian.net/browse/TMEDIA-252)

## Acceptance Criteria
Expected Behavior
1. No additional scrolling should happen when you reach the bottom of a page when
a. The sections menu is closed
b. The sections menu is closed and have more items than would fit within the viewport
2. When the sections menu is open
a. The sections menu should scroll to allow a user to see all menu items
b. User should be able to open multiple sub sections and still scroll to see all menu items

## Test Steps
- Add test steps a reviewer must complete to test this PR

1. Checkout this branch `git checkout TMEDIA-252-nav-list-whitespace-issue`
2. Run fusion repo with linked blocks `npx fusion start -f -l @wpmedia/shared-styles,@wpmedia/header-nav-chain-block`
3. Go to http://localhost/homepage/?_website=the-gazette 
4. Click sidebar section
5. Open subsection
6. Scroll down the page and ensure background color of items is not transparent

## Effect Of Changes
### Before
![Screen Shot 2021-07-27 at 11 58 00](https://user-images.githubusercontent.com/5950956/127196124-a9c3be11-cad5-4042-80b2-0f24541edb7e.png)

### After
![Screen Shot 2021-07-27 at 11 57 25](https://user-images.githubusercontent.com/5950956/127196046-195fe278-6ff8-4b7b-85ad-51c358b200e1.png)

## Dependencies or Side Effects

## Author Checklist
_The author of the PR should fill out the following sections to ensure this PR is ready for review._
- [x] Confirmed all the test steps a reviewer will follow above are working.
- [x] Confirmed there are no linter errors. Please run `npm run lint` to check for errors. Often, `npm run lint:fix` will fix those errors and warnings.
- [x] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [x] Confirmed this PR has reasonable code coverage. You can run `npm run test:coverage` to see your progress.
  - [ ] Confirmed this PR has unit test files
  - [ ] Ran `npm run test`, made sure all tests are passing
  - [ ] If the amount of work to write unit tests for this change are excessive,
please explain why (so that we can fix it whenever it gets refactored).
- [ ] Confirmed relevant documentation has been updated/added.

## Reviewer Checklist
_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] Linting code actions have passed.
- [ ] Ran the code locally based on the test instructions.
  - [ ] I don’t think this is needed to be tested locally. For example, a padding style change (storybook?) or a logic change (write a test).
- [ ] I am a member of the engine theme team so that I can approve and merge this. If you're not on the team, you won't have access to approve and merge this pr.
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.

Commit:
fix: Add background color fallback essentially for list items in case off-screen initially

nschubach commented:
Is this supposed to cover the issue of cut off content on the menu too?

![image](https://user-images.githubusercontent.com/2287238/127202680-50a5cd07-283d-4f69-8980-731c3e06c56f.png)

![image](https://user-images.githubusercontent.com/2287238/127202713-7acb1cdf-e971-40d6-82bd-652efbd075fb.png)

nschubach commented:
I briefly looked into that and since we are setting the height to 100vh then shifting the margin down, the content is all showing, but it's shifted down by the size of the header bar.  The way I got it to show was to do the following based on the margin:
![image](https://user-images.githubusercontent.com/2287238/127204120-3ac13816-9662-4d3b-a366-4b6cb61568f8.png)

JackHowa commented:
hmm interesting. I can look into that cutoff issue. seems near in scope at least @nschubach

Commit:
fix: Accomodate for top nav height and margin based on feedback

Via https://github.com/WPMedia/fusion-news-theme-blocks/pull/1010#issuecomment-887714729

Commit:
fix: Ensure mobile nav working as expected as well taking into account hard-coded values

Commit:
Merge branch 'canary' into TMEDIA-252-nav-list-whitespace-issue

JackHowa commented:
updated -- big thanks @nschubach 

# before 
- last items were being hidden (brand studio) or unstyled on scroll
![image](https://user-images.githubusercontent.com/5950956/127211609-2cb4217f-4c46-442b-a132-ceb0540a49e5.png)
![Screen Shot 2021-07-27 at 13 57 23](https://user-images.githubusercontent.com/5950956/127211900-9fdbe1b4-c64d-4afc-81dc-23505892efb0.png)

# after

## desktop 

- accommodating on desktop and mobile for margins and items above with hard-coded values 

## mobile 

![Screen Shot 2021-07-27 at 13 51 17](https://user-images.githubusercontent.com/5950956/127211682-25be070a-58a0-4e09-aa9a-7765c7078f9f.png)
![Screen Shot 2021-07-27 at 13 51 10](https://user-images.githubusercontent.com/5950956/127211693-669e812e-d289-48b0-9314-e2637c601676.png)

matthewroach commented on blocks/shared-styles/scss/_header-nav.scss:
```diff
@@ -408,17 +409,27 @@ body.nav-open {
     background: rgba(25, 25, 25, 0.5);
-    height: 100vh;
+    // total viewport height - header height
+    height: calc(100vh - 100px);
```
What's the `100px` value? This is a magic number as the header height is controllable as a custom field

nschubach replied:
That 100px is the margin that is applied to the top of the menu and is basically the header height.  Unfortunately, it's not easy to expose that to the stylesheet where that is needed I don't think.  I have been playing around with options to counter this but have no real solutions that don't up-end the page a bit more than comfort allows.

JackHowa replied:
I've updated this for styled components and custom fields

matthewroach commented on blocks/shared-styles/scss/_header-nav.scss:
```diff
@@ -408,17 +409,27 @@ body.nav-open {
     .section-menu {
-      height: 100vh;
+      // total viewport height - header height - top margin style
+      height: calc(100vh - 100px - 13px);
```
Again, magic numbers are a little scary with possible changes

matthewroach commented on blocks/shared-styles/scss/_header-nav.scss:
```diff
@@ -408,17 +409,27 @@ body.nav-open {
+
+      .section-menu {
+        // viewport height - search sub nav height - top nav height
+        height: calc(100vh - 48px - 56px);
```
More magic numbers?

JackHowa replied:
updated -- that's my bad

matthewroach reviewed (changes requested):
Seeing a lot of white space at the bottom of the page in FireFox still

![firefox-white-space-end-of-page](https://user-images.githubusercontent.com/868127/127220041-1ef491c9-94d0-4032-9236-4c69e2dc6e25.gif)

JackHowa commented:
ok interesting I'll have to maybe look at refactoring this more @matthewroach

Commit:
fix: Remove magic numbers, remove hard-coded top margin. Take in dynamic height

Commit:
test: Change for styled components inclusion, still rendering classes

Commit:
Merge branch 'TMEDIA-252-nav-list-whitespace-issue' of github.com:WPMedia/fusion-news-theme-blocks into TMEDIA-252-nav-list-whitespace-issue

JackHowa commented:
I removed the hard-coded widths and heights from the stylesheet and am doing it dynamically based on custom fields. So this is an improvement; I did not realize that the height of the nav was variable. that's my bad. I did remove the top margin 13 px that seemed like it was for style and did not want to hard-code into the styled component. 

with scroll down 

<img width="721" alt="Screen Shot 2021-07-27 at 16 47 13" src="https://user-images.githubusercontent.com/5950956/127231832-2e82a828-0fd5-4708-8f9c-e422e8d9ec85.png">

without scroll down 

<img width="689" alt="Screen Shot 2021-07-27 at 16 47 37" src="https://user-images.githubusercontent.com/5950956/127231888-714308f6-7ebf-4d17-8525-340f408f499d.png">

JackHowa commented:
The firefox bug seems to be common based on other reports and their bugtracker. the overflow property does not seem to be working as it does in chromium browsers 

![Screen Shot 2021-07-27 at 16 17 13](https://user-images.githubusercontent.com/5950956/127232123-2d81ded7-b0d2-49ef-b9bc-00ec98c9a86f.png)


https://stackoverflow.com/questions/28636832/firefox-overflow-y-not-working-with-nested-flexbox/28639686#28639686
![Screen Shot 2021-07-27 at 16 49 35](https://user-images.githubusercontent.com/5950956/127232061-deeb47e5-dc6b-44e7-9665-a6f7894de356.png)


I think this solves the bug I reported of the bottom list items not being able to be styled as expected. I'm not too sure what our support for firefox should be here.

Commit:
style: Run stylelint fix for newline fixes

matthewroach commented:
@JackHowa - The original report was due based on seeing the issue in Firefox, and we have had reports and fixed related issues in the past that have affected on FireFox - https://arcpublishing.atlassian.net/browse/TMEDIA-174.
The way the section menu works, does have some quirks cross browser in it's current implementation which we noted during grooming - it's very possible it needs refactored to a better set up as a whole

nschubach commented:
I just pulled down the latest and ran it (changed the run command to add the header block: `npx fusion start -f -l @wpmedia/engine-theme-sdk,@wpmedia/header-nav-chain-block`) and it's now doing this:

![image](https://user-images.githubusercontent.com/2287238/127337599-beb90a44-00f8-44af-a708-bf73c86f6ae6.png)

I'm with Matt here.  There is a structure issue here.

Commit:
fix: Hide overflow when closed to ensure window in firefox doesn't grow vertically

Commit:
style: Remove duplicate selector

Commit:
style: Fix newline stylelint

Commit:
Merge branch 'canary' into TMEDIA-252-nav-list-whitespace-issue

JackHowa commented:
@matthewroach updated per firefox fix 


before 

- scroll down the page on firefox https://corecomponents-the-gazette-prod.cdn.arcpublishing.com/ 

<img width="1439" alt="Screen Shot 2021-07-28 at 11 00 32" src="https://user-images.githubusercontent.com/5950956/127356381-d8609b34-e6fa-4056-a7b4-aefce7f0393a.png">


after

- go to  http://localhost/homepage/?_website=the-gazette
- scroll down the page. validate that it doesn't continue scrolling past the footer 
- play with the sidebar nav. ensure that expanding and closing sites does overflow and make the page longer

<img width="1440" alt="Screen Shot 2021-07-28 at 10 53 46" src="https://user-images.githubusercontent.com/5950956/127356678-8938f272-f7ef-478d-81f9-2ad3276bfcab.png">

JackHowa commented:
@matthewroach updated per firefox fix -> good to go


before 

- scroll down the page on firefox https://corecomponents-the-gazette-prod.cdn.arcpublishing.com/ 

<img width="1439" alt="Screen Shot 2021-07-28 at 11 00 32" src="https://user-images.githubusercontent.com/5950956/127356381-d8609b34-e6fa-4056-a7b4-aefce7f0393a.png">


after

- go to  http://localhost/homepage/?_website=the-gazette
- scroll down the page. validate that it doesn't continue scrolling past the footer 
- play with the sidebar nav. ensure that expanding and closing sites does overflow and make the page longer

<img width="1440" alt="Screen Shot 2021-07-28 at 10 53 46" src="https://user-images.githubusercontent.com/5950956/127356678-8938f272-f7ef-478d-81f9-2ad3276bfcab.png">

Commit:
fix: Add back top nav list stylistic margin

Commit:
Merge branch 'TMEDIA-252-nav-list-whitespace-issue' of github.com:WPMedia/fusion-news-theme-blocks into TMEDIA-252-nav-list-whitespace-issue

Commit:
fix: Remove test code for testing safari fixes for bottom nav

JackHowa commented:
@matthewroach added back the top-margin that we talked about. 

- I've been testing with my actual phone and the simulators and haven't come to a good solution yet for the bottom margin
- I forgot the safe area inset doesn't seem accommodate for the actual safari bottom toolbar. but I've had some success with the margin-bottom on the nav-section for mobile breakpoints

Commit:
fix: Ensure bottom link is visible on mobile phones and on desktop on mouseover with placeholder div

JackHowa commented:
Ensure bottom link is always visible 

# before 

## desktop 

![Screen Shot 2021-07-29 at 10 10 45](https://user-images.githubusercontent.com/5950956/127517366-f339b3c0-cd5c-4ec7-9659-9f34065b9fbc.png)

## mobile 

<img width="371" alt="Screen Shot 2021-07-29 at 10 10 15" src="https://user-images.githubusercontent.com/5950956/127517283-4667f215-ba78-4af8-8007-e4c626a1de15.png">

# after

## desktop 

![Screen Shot 2021-07-29 at 10 07 48](https://user-images.githubusercontent.com/5950956/127517023-0679dad7-1e40-4082-a237-2a3fb4291727.png)

## mobile 

<img width="391" alt="Screen Shot 2021-07-29 at 10 08 57" src="https://user-images.githubusercontent.com/5950956/127517110-29e7eda9-cc4e-4b72-9047-9902e9472022.png">

Commit:
style: LInt styles fix for order

Commit:
style: Use variable for background color to fix list items showing up transparent

Commit:
style: Remove unnecessary newline

nschubach commented:
I'm still getting this weirdness when making the window really short and scrolling down in the menu:
![image](https://user-images.githubusercontent.com/2287238/127550094-0acd46c0-ebe3-4ca4-b654-11dcbe442199.png)

JackHowa commented:
I'm not able to replicate this @nschubach 

what blocks are you linking? should be: 

`npx fusion start -f -l @wpmedia/shared-styles,@wpmedia/header-nav-chain-block`

Also I'd make sure that the branch is updated. A lot has changed since this was started

<img width="1263" alt="Screen Shot 2021-07-29 at 15 26 33" src="https://user-images.githubusercontent.com/5950956/127561035-299c6f66-b358-43ba-b0ec-2f7fe550e56f.png">

nschubach commented:
I had the wrong link 🤦 .  looks good now.
